panel/panelpluginsmodel.cpp

```diff
@@ -107,11 +107,23 @@ QList<Plugin *> PanelPluginsModel::plugins() const
     return std::move(plugins);
 }
 
+Plugin* PanelPluginsModel::pluginByName(QString name) const
+{
+    for (auto const & p : mPlugins)
+        if (p.first == name)
+            return p.second.data();
+    return nullptr;
+}
+
 void PanelPluginsModel::addPlugin(const LxQt::PluginInfo &desktopFile)
 {
     QString name = findNewPluginSettingsGroup(desktopFile.id());
     QPointer<Plugin> plugin = loadPlugin(desktopFile, name);
-    if (!plugin.isNull())
+    if (plugin.isNull())
+        return;
+
+    bool single = plugin->iPlugin()->flags().testFlag(ILxQtPanelPlugin::SingleInstance);
+    if (!single || !pluginNames().contains(desktopFile.id()))
```

Comment:
>> Does this even work for you? Here the 'type' is searched plugins names (instances) => can't be found there.

>I don't follow you here. But it does work for me.

`pluginNames()` are instances -> e.g: 'mainmenu2', 'spacer2', 'spacer3', 'tray2' etc.. in contrast with `desktopFile.id()` which is base name of the .desktop file -> e.g.: 'mainmenu', 'spacer', 'tray'

You are just lucky that your 'tray' plugin instance doesn't have autogenerated name. Try to remove the plugin from panel and reinsert it after that.


>> Actualy while reviewing this I think that the SingleInstance flag should be known before constructing the Plugin object => somewhere in ILxQtPanelPluginLibrary to not create next object if one already created.

> That would be great, but I don't know how to do it.

It must be known somewhere here https://github.com/lxde/lxqt-panel/blob/master/panel/plugin.cpp#L251 => so maybe implemented directly in `instance()` method ... that means specific plugin implementation as also @jleclanche suggested.